components/PopulationCalculation.tsx

```diff
@@ -0,0 +1,137 @@
+import { showNotification } from '@mantine/notifications';
+import { IconAlertCircle, IconX } from '@tabler/icons';
+
+export default function PopulationCalculation() {
+  const currentPatients = useRecoilValue(patientTestCaseState);
+  const measureBundle = useRecoilValue(measureBundleState);
+  const measurementPeriod = useRecoilValue(measurementPeriodState);
+  const [measureReports, setMeasureReports] = useState<DetailedMeasureReport[]>([]);
+  const [opened, setOpened] = useState(false);
+
+  /**
+   * Creates array containing patient info strings for each created patient. Used as labels when creating
+   * measure report population results table.
+   * @returns { Array } an array of patient labels
+   */
+  const createPatientLabels = (): string[] => {
+    const patientLabels: string[] = [];
+    Object.keys(currentPatients).forEach(id => {
+      patientLabels.push(getPatientInfoString(currentPatients[id].patient));
+    });
+    return patientLabels;
+  };
+
+  /**
+   * Uses fqm-execution library to perform calculation on all patients and return their
+   * measure reports.
+   * @returns { Array | void } array of measure reports (if measure bundle is provided)
+   */
+  const calculateMeasureReports = async (): Promise<fhir4.MeasureReport[] | void> => {
+    // specify options for calculation
+    // TODO: revisit options after new fqm-execution release (calculateSDEs set to true throws error)
+    const options: CalculatorTypes.CalculationOptions = {
+      calculateHTML: false,
+      calculateSDEs: false,
+      reportType: 'individual',
+      measurementPeriodStart: measurementPeriod.start?.toISOString(),
+      measurementPeriodEnd: measurementPeriod.end?.toISOString()
+    };
+
+    // get all patient bundles as array to feed into fqm-execution
+    const patientBundles: fhir4.Bundle[] = [];
+    Object.keys(currentPatients).forEach(id => {
+      const bundle = createPatientBundle(currentPatients[id].patient, currentPatients[id].resources);
+      patientBundles.push(bundle);
+    });
+
+    if (measureBundle.content) {
+      const measureReports = await Calculator.calculateMeasureReports(measureBundle.content, patientBundles, options);
+      return measureReports.results as fhir4.MeasureReport[];
+    } else return;
+  };
+
+  /**
+   * Wrapper function that calls calculateMeasureReports() and creates the DetailedMeasureReport that will be used to render
+   * the population results. Catches errors in fqm-execution that result from calculateMeasureReports().
+   */
+  const runCalculation = () => {
+    calculateMeasureReports()
+      .then(measureReports => {
+        if (measureReports) {
+          const patientLabels = createPatientLabels();
+          const labeledMeasureReports: DetailedMeasureReport[] = [];
+          measureReports.forEach((mr, i) => {
+            labeledMeasureReports.push({ label: patientLabels[i], report: mr as fhirJson.MeasureReport });
```

Comment:
Can we always rely on the measureReports for each patient being returned in the same order that the patients are in the passed-in bundle? It seems like that is what this logic is relying on

Reply:
prob shouldn't. Should look at the mr subject to ensure that patient ID matches